feat(admin): workspace apps

allardy opened the pull request:
This PR is kind of big because I had to change a couple of things to display modules on the admin panel.

Copied the content from `botpress/ui` to ModuleUI in `botpress/shared` so modules no longer depend on the studio. We can't remove it from the studio yet because i'm pretty sure they are used by custom modules.

## Modules
- Adds the 'workspaceApp' field to module definition to display apps on the admin panel. Users can bookmark the URL to open an app directly on a specific bot, ex: http://localhost:3000/admin/apps/hitlnext/welcome-bot
- An app can be global (displayed on the left menu) or for bots (displayed in the bots submenu). It could also be both... For example the code editor (which isn't, because we don't intend to do modifications in production)

![image](https://user-images.githubusercontent.com/42552874/116570411-417c4e80-a8d8-11eb-95b3-6cb9aee8c66e.png)


## Admin
- Copied the InjectedModuleView logic from the studio so modules can be displayed correctly
- Added the 'ConfigStatus' component to restart the server on botpress config changes
- Added a Language switcher when using a workspace app (also displays the name of the bot)
- Fix some permissions issues with the display (menu icons displayed wrongly, error messages because of lack of permissions, etc

![image](https://user-images.githubusercontent.com/42552874/116570593-640e6780-a8d8-11eb-9426-394e5d407634.png)



## Webchat
- Add 'loadConversation' event to the webchat so HITL doesn't need the store
- Fix decoding issue when HITL loads the webchat on the admin panel (the URI is already decoded)

## HITL Next
- Refactored a bit of HITL LiveChat to ensure the webchat is fully loaded & doesn't need the webchat store

## Code Editor
- The module is available on the admin panel and displays global files by default. The advanced editor is also available

![image](https://user-images.githubusercontent.com/42552874/116576743-d5044e00-a8dd-11eb-8861-20f12f4f4dab.png)

## Misunderstood
- To avoid adding the material icons on the admin panel only for this icon, i've updated the icon logic slightly
- Small change to the ElementPreview element so it doesn't depend on the studio

allardy committed:
chore(dev): mini fixes on admin

allardy committed:
update styling

allardy committed:
fix cmd

allardy committed:
minor change

allardy committed:
feat(admin): load modules view

allardy committed:
fix assets and ws

allardy committed:
fix paths

allardy committed:
fix login

allardy committed:
fix test

allardy committed:
Merge branch 'ya-admin-base' into ya-admin-modules

allardy committed:
fix hitl and web

allardy committed:
studio redirect and test

allardy committed:
Merge branch 'ya-admin-base' into ya-admin-modules

allardy committed:
fix style

allardy committed:
chore(dev): better ui for admin panel

allardy committed:
pr comments

allardy committed:
feat(admin): move realtime logs to admin

allardy committed:
add bot filter

allardy committed:
remove test for deleted page

allardy committed:
fix error

allardy committed:
Merge branch 'master' into ya-admin-logs

allardy committed:
Merge branch 'master' into ya-admin-modules

allardy committed:
Merge branch 'ya-admin-logs' into ya-admin-modules

allardy committed:
minor changes

allardy committed:
add bot name and language selector

allardy committed:
mini fixes

allardy committed:
fix misunderstood icon

allardy committed:
Merge branch 'master' into ya-admin-modules

allardy committed:
fix element preview

allardy committed:
fix logo

allardy committed:
fix logo

allardy committed:
update libraries

allardy committed:
add deprecate notice

allardy committed:
add bot-scoped and global workspace apps

allardy committed:
fix menu

allardy committed:
fix permissions

allardy committed:
fix ui

allardy committed:
cleanup

allardy committed:
fix scrollbar

allardy committed:
add tests

allardy committed:
fix translations

allardy committed:
fix path

allardy committed:
fix misunderstood categories & menu item

allardy commented on modules/hitlnext/src/views/full/app/components/LiveChat.tsx:
```diff
@@ -11,32 +11,59 @@ interface Props {
+    window.document.body.appendChild(script)
+
+    intervalId = setInterval(() => {
+      if (window.botpressWebChat) {
```
We must ensure that the webchat is loaded only once, but even if the script tag is added, it doesn't means that the webchat is ready. The first check ensures that the script is loaded, and the second ensures that it has loaded completely. Since the method is only called once, there should not be any issues

laurentlp replied:
Oh okay, I see! Thanks for the clarification. Can you please add your response as a comment in the code so that it will also be clear for everyone?

allardy committed:
pr comments

allardy commented on src/bp/admin/ui/src/app/StatusBar/LangSwitcher.tsx:
```diff
@@ -0,0 +1,77 @@
+    const flag = require(`../../management/languages/flags/${code}.svg`)
+    return flag.default || flag
+  } catch (err) {
+    return requireFlag('missing')
```
I wanted to avoid the double try catch, but I've added it

laurentlp replied:
Why not simply do:

```ts
const requireFlag = code => {
  try {
    const flag = require(`../../management/languages/flags/${code}.svg`)
    return flag.default || flag
  } catch (err) {
    if (code === 'missing') {
       // return || throw
    }
    return requireFlag('missing')
  }
}
```

allardy replied:
Good point :)

allardy commented on src/bp/core/security/router-security.ts:
```diff
@@ -229,7 +230,7 @@ export const checkBotVisibility = (configProvider: ConfigProvider, checkTokenHea
   // '___' is a non-valid botId, but here acts as for "all bots"
   // This is used in modules when they setup routes that work on a global level (they are not tied to a specific bot)
   // Check the 'sso-login' module for an example
```
Good point

allardy committed:
pr comments

laurentlp commented on src/bp/admin/ui/src/app/StatusBar/LangSwitcher.tsx:
```diff
@@ -0,0 +1,81 @@
+
+  const restoreLastLanguage = () => {
+    const lastLang = localStorage.getItem(STORAGE_KEY)
+    if (!props.languages || !props.languages.length || !lastLang) {
```
isn't optional chaining available through the admin?

allardy replied:
It does. Just did a lot of copy and paste without reviewing stuff

laurentlp committed:
Merge branch 'master' into ya-admin-modules

laurentlp commented:
> ## Admin
>  - ...
> - Added a Language switcher when using a workspace app (also displays the name of the bot)

@allardy Not sure I understand how this can be used? Or even why this is useful?

allardy commented:
> > ## Admin
> > 
> > * ...
> > * Added a Language switcher when using a workspace app (also displays the name of the bot)
> 
> @allardy Not sure I understand how this can be used? Or even why this is useful?

I had to add it so misunderstood was able to work correctly. On the studio, it depends on the lang switcher to display misunderstood to users. Some modules "might" need it, so I've added it for workspace apps. Maybe we may include it in the module directly in the future, but i'm trying to keep modules working on the admin and studio for the time being

allardy committed:
pr comments

laurentlp commented:
@allardy I don't know if this is the expected behavior, but each time I go to the studio (with a Content Editor account), I see those errors:

![Screenshot from 2021-05-05 17-54-48](https://user-images.githubusercontent.com/9640576/117214662-355a2a80-adcb-11eb-9760-99d040e3578d.png)

I know that the Content Editor does not have the rights to edit flows, but shouldn't we lock it instead of raising errors?

laurentlp commented:
@allardy I think we should either remove or uncheck by default the `Create for the current bot` option when adding a new action or hook from the 'admin' Code Editor. This would make more sense since the 'admin' code editor bot is the '__' one.

Also, when you're a Developer, you cannot seem to be able to create global actions/hooks. What do you think of simply hiding this feature for Developers too?

allardy committed:
fix permissions

allardy commented:
Those are all valid observations, thanks, I've fixed them. 

Ideally i'd like to hide the code editor from the admin when the user doesn't have access to global stuff, but only the module "knows" what content it offers and if the user has access, since the menu only checks if the user "has access to the module" (and the user does, but only for bots-scoped stuff)

Also, somehow the readOnly flag was not enabled properly, i've fixed that for content editor. Fortunately, it only impacts front-end, not backend logic

allardy committed:
Merge branch 'master' into ya-admin-modules

sebburon commented:
> @allardy I don't know if this is the expected behavior, but each time I go to the studio (with a Content Editor account), I see those errors:
> 
> ![Screenshot from 2021-05-05 17-54-48](https://user-images.githubusercontent.com/9640576/117214662-355a2a80-adcb-11eb-9760-99d040e3578d.png)
> 
> I know that the Content Editor does not have the rights to edit flows, but shouldn't we lock it instead of raising errors?

Just pointing out that this behaviour is also present on the agent role of HITLnext

laurentlp commented on src/bp/ui-studio/src/web/views/FlowBuilder/index.tsx:
```diff
@@ -102,8 +102,10 @@ const FlowBuilder = (props: Props) => {
 
-    setReadOnly(false)
-    setMutex(undefined)
+    if (isOperationAllowed({ operation: 'write', resource: 'bot.flows' })) {
+      setReadOnly(false)
+      setMutex(undefined)
+    }
```
Somehow, I have to refresh the page at least one in order for the Flow Editor to be locked:


https://user-images.githubusercontent.com/9640576/117307106-48faa500-ae4e-11eb-8f0c-f04e22540a43.mp4

allardy commented:
@sebburon That's a good point, i'll see how we can simply remove all access to the studio for HITL agents, since the app is now on the admin panel

@laurentlp i'm not able to reproduce, everytime I start the server and try to open the flow, it's locked. Even tried to switch account and come back

allardy committed:
hide studio for hitl agents

laurentlp commented:
> @laurentlp i'm not able to reproduce, everytime I start the server and try to open the flow, it's locked. Even tried to switch account and come back

Yeah, my setup was kind of unusual. I tested it once more and it works fine!

allardy committed:
fix editor permissions and translation

laurentlp reviewed (approved):
Good job!!! :100:
